posts/2018-02-09-serverless-and-graphql-in-love.md

```diff
@@ -0,0 +1,611 @@
+
+![](https://user-images.githubusercontent.com/1587005/36131686-dac3658a-1027-11e8-9d5d-4a435f5ae43b.png)
+ 
+#### Cold Starts:
+
+Approximately, 2% of the total calls were [cold starts]((https://hackernoon.com/im-afraid-you-re-thinking-about-aws-lambda-cold-starts-all-wrong-7d907f278a4f) ), where I noticed an additional latency of 700ms - 800ms in lambda execution time for the first API call which came from initialization of the lambda container itself. 
+
+This additional latency was observed in both endpoints, i.e., with DynamoDB and PostgreSQL. There are ways to optimize this overhead, and I would strongly recommend you to read [this post](https://serverless.com/blog/keep-your-lambdas-warm/) for the same
+ 
+#### Increase in Lambda memory size limit by 2x and 3x:
+
+Increasing the lambda memory size by 2x (2048 MB) improved the overall latency of lambda service by 18% and 3x (3008 MB) improved the latency by 38%. The latency of DynamoDB backend remained constant and the lambda execution time itself improved within 20% range for 3x memory.  
+
+Lambda Service Latency (1GB Memory)        |  Lambda Service Latency (2GB Memory)
+:-------------------------:|:-------------------------:
+![](https://user-images.githubusercontent.com/1587005/36127284-87ae8f16-1011-11e8-9f9a-d1435a066d06.png)  |  ![](https://user-images.githubusercontent.com/1587005/36127303-959b9db2-1011-11e8-9fd4-1d9556a6dc25.png)
+
+#### Lambda execution time with PostgreSQL backend:
+
+In case of RDS, lambda execution time increases with increase in the size of the data. For instance, when I repeated my analysis with an increase in Tweets dataset by a factor of 100 (i.e. 1000 tweets per user) and I found the response time increased by 5x - 10x. This possibly happens because we are joining Tweets and Users table on the fly which results in more query execution time. Query performance can be further improved by using indexing and other database optimizations (not in the scope of this blog). On the other hand, latency of DynamoDB remains constant with increase in dataset size (which is expected by design).
+
+#### API Gateway and Network Latency:
+
+On average, the E2E response time of the GraphQL endpoint ranges from 100ms - 200ms (including the lambda execution time). Hence on API Gateway and network latency is approximately between 40 - 100 ms, which can be further reduced by caching. You might ask - Why do we need API Gateway? Can we merely use lambda to fetch the GraphQL response? I will delve more into this in my future blogs, but for now, you can read [this post](https://forum.serverless.com/t/convince-me-to-use-api-gateway-and-not-call-lambda-direct/3214) which evaluates the same. 
+
+**Note**:  This analysis merits a separate blog of its own where we can do an in-depth study of all the latencies and query optimizations. (Coming Soon!) 
+
+## Selling GraphQL in your organization
```

Comment:
That's a really interesting and important topic IMHO. Would be worthwhile do add a note about that in the introduction :+1: